src/gui/qgsmaptool.h

```diff
@@ -205,14 +207,16 @@ class GUI_EXPORT QgsMapTool : public QObject
      * \a menu will be initially populated with a set of default, generic actions.
      * Any new actions added to the menu should be correctly parented to \a menu.
      *
+     * A pointer to the map mouse \a event can be provided to allow particular behavior depending on the map tool.
+     *
      * The default implementation does nothing.
      *
      * \note The context menu is only shown when the ShowContextMenu flag
      * is present in flags().
      *
      * \since QGIS 3.14
      */
-    virtual void populateContextMenu( QMenu *menu );
+    virtual void populateContextMenu( QMenu *menu, QgsMapMouseEvent *event = nullptr );
```

Comment:
This is effectively an API change, any plugin implementing this method will no longer work after this change.

@nyalldawson how do we handle such things?

Reply:
The logic is now:
```
  if ( mMapTool )
    if ( !mapTool()->populateContextMenuWithEvent( &menu, event ) )
      mMapTool->populateContextMenu( &menu );
```

`populateContextMenuWithEvent( )` return false if not implemented.